### ISSUE TITLE
Enable proxy protocol for Ingress Envoy

### ISSUE THREAD

**Issue by b1tamara** [cloudfoundry/cf-for-k8s#561]
**Is your feature request related to a problem? Please describe.**
Some Load Balancers like AWS ELB replace the client IP by their own IP. In that case the feature "X-Forwarded-For" cannot be used because the X-Forwarded-For-Header contains the IP of the Load Balancer. By using Proxy Protocol between LB and Ingress Envoy the client IP will be set in the X-Forwarded-For-Header by Envoy.

**Describe the solution you'd like**
Apply EnvoyFilter for enabling the proxy protocol on the Envoy side. Make it configurable via values.  

```
#@ load("@ytt:data", "data")
#@ if data.values.enable_proxy_protocol:
---
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: proxy-protocol
  namespace: istio-system
spec:
  workloadSelector:
    labels:
      istio: ingressgateway
  configPatches:
  - applyTo: LISTENER
    patch:
      operation: MERGE
      value:
        listener_filters:
        - name: envoy.listener.proxy_protocol
        - name: envoy.listener.tls_inspector
#@ end
```
**Additional context**
The enabling of the proxy protocol of the Load Balancer side is not part of this feature request because it is a Load Balancer specific configuration.

**Comment by mike1808** [comment]
Hi. Can you say why `X-Forwarded-For` header is not a solution? AFAIK Envoy will just add another IP address to that header so it will contain 1. ip of the user 2. ip of the AWS ELB

**Comment by mike1808** [comment]
Also, have you tried configuring `gatewayTopology` config in Istio? https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/#configuring-x-forwarded-for-headers

**Comment by b1tamara** [comment]
Hello mike1808,
thank you for your suggested solution. Unfortunately it will not work for LoadBalancers which are not transparent (Transparent indicates whether client IP addresses are visible to the backend servers). 
I enabled the proxy protocol on AWS LB that the Load Balancer preserves the client IP. Then I applied the IstioOperator using the annotation for Gateway istio-ingressgateway:
```
"proxy.istio.io/config": "{\"gatewayTopology\":{\"numTrustedProxies\":1}}",
```
After that I tried to curl the app which returns the "x-Forwarded-For-Header". I got:
```
curl https://x-forwarded-for.<domain>

curl: (35) LibreSSL SSL_connect: SSL_ERROR_SYSCALL in connection to x-forwarded-for.<domain>
```
The problem is that the LoadBalancer tries to speak with EnvoyProxy the proxy protocol which Envoy does not understand as it is disabled per default. 

If I apply the EnvoyFilter in order to enable the proxy protocol on the Envoy side, I got: 
```
curl https://x-forwarded-for.<domain>
<h1>X-Forwarded-For: <myClientIP></h1>  
```
We would like to provide the PR in order to enable this feature. 

Best regards,
Tamara

**Comment by shalako** [comment]
I heard about this PR and wanted to thank the SAP team for contributing!

With regard to user experience for enabling this capability in cf-for-k8s, what is the risk of having it enabled all the time, rather than offering a field in `cf-values.yml` to turn it on and off?

**Comment by stefanlay** [comment]
Hi @shalako, thanks for the feedback.

The proxy protocol needs to be enabled on both sides of the communication. The proxy_protocol filter in Envoy expects that the proxy protocol is used, otherwise the tcp connection cannot be established.

We also tested it with AWS and GCP loadbalancers. In both cases we got the error:

```
wget: error getting response: Connection reset by peer
Connecting to <URL> (<IP>:443)
ssl_client: <URL>: handshake failed: Connection reset by peer
wget: error getting response: Connection reset by peer
```

Therefore we think that it is necessary to make it configurable.
